File: tests/test_file_chip.py
"""The file chip's tool list is the server's file-writing tools, nothing else.

Live 2026-09-24: x_list_scheduled -- a read -- showed "WROTE Asia/Kuwait".
chat.js now shows the chip only for tools in _FILE_CHIP_OPS. This keeps that
list honest against the single source of truth for what a tool does:
kazma_core.safety.side_effects (EffectKind.WRITE_FS). Behaviour:
tests/js/test_file_chip.js.
"""

from __future__ import annotations

import re
from pathlib import Path

from kazma_core.safety.side_effects import EffectKind, get_effect_profile

CHAT_JS = Path(__file__).resolve().parent.parent / "kazma-ui" / "kazma_ui" / "static" / "js" / "chat.js"
CORE_FILE_TOOLS = {"file_write", "file_append", "file_apply_patch", "file_apply_patch_set", "file_delete"}


def _chip_tools(src: str) -> set[str]:
    block = re.search(r"var _FILE_CHIP_OPS = \{(.*?)\};", src, re.DOTALL)
    assert block, "_FILE_CHIP_OPS not found in chat.js"
    return set(re.findall(r"^\s*(\w+)\s*:", block.group(1), re.MULTILINE))


def _not_file_writers(tools: set[str]) -> list[str]:
    return sorted(t for t in tools if get_effect_profile(t).effect is not EffectKind.WRITE_FS)


def test_every_chip_tool_writes_files():
    tools = _chip_tools(CHAT_JS.read_text(encoding="utf-8"))
    assert not _not_file_writers(tools), (
        "a chip that says WROTE for a tool the side-effect registry does not "
        f"classify WRITE_FS: {_not_file_writers(tools)}"
    )


def test_the_core_file_tools_keep_their_chip():
    tools = _chip_tools(CHAT_JS.read_text(encoding="utf-8"))
    assert CORE_FILE_TOOLS <= tools, sorted(CORE_FILE_TOOLS - tools)


def test_the_parity_gate_catches_a_read_tool():
    """Negative control (AGENTS.md section 28): the live mislabel."""
    bad = "var _FILE_CHIP_OPS = {\n  file_write: 'wrote',\n  x_list_scheduled: 'wrote'\n};"
    assert _not_file_writers(_chip_tools(bad)) == ["x_list_scheduled"]

File: tests/js/test_file_chip.js
/**
 * The workbench file chip ("WROTE path") appears only for tools that write
 * files, and names the file they wrote.
 *
 * Live 2026-09-24: x_list_scheduled -- a read -- showed "WROTE Asia/Kuwait".
 * The chip guessed from the tool name and from any slash in the result, and
 * its output held a timezone. Runs the REAL _FILE_CHIP_OPS / _fileChipOp /
 * _extractPathFromTool from chat.js. Parity with the server's side-effect
 * registry: tests/test_file_chip.py.
 */
"use strict";

const fs = require("fs");
const path = require("path");

const src = fs.readFileSync(path.join(
  __dirname, "..", "..", "kazma-ui", "kazma_ui", "static", "js", "chat.js"), "utf8")
  .replace(/\r\n/g, "\n");
const start = src.indexOf("  var _FILE_CHIP_OPS = {");
const end = src.indexOf("  function _renderPlanList(", start);

let fail = 0;
function ok(name, cond, detail) {
  if (cond) { console.log("OK " + name); return; }
  fail += 1;
  console.log("FAIL " + name + "  " + String(detail || ""));
}
ok("chip functions are extractable", start > 0 && end > start);

// eslint-disable-next-line no-new-func
const api = new Function(src.slice(start, end) +
  "\nreturn { op: _fileChipOp, path: _extractPathFromTool };")();

const live = '{"ok": true, "count": 12, "posts": [{"id": 12, "tz": "Asia/Kuwait", ' +
  '"fire_at": "2026-09-03T17:19:26+03:00"}]}';
ok("live: a read with a timezone in its output gets no chip",
  api.path("x_list_scheduled", live) === "", api.path("x_list_scheduled", live));
ok("a shell result with a path gets no chip",
  api.path("shell_exec", "ls /usr/local/bin") === "");
ok("a fetched URL gets no chip",
  api.path("read_url", "https://example.com/a/b") === "");
ok("x_status gets no chip", api.op("x_status") === "");

ok("file_write names the file from its result",
  api.path("file_write", "Wrote 3 lines, 20 bytes to docs/notes.md") === "docs/notes.md",
  api.path("file_write", "Wrote 3 lines, 20 bytes to docs/notes.md"));
ok("file_write names the file from its arguments",
  api.path("file_write", '{"path": "src/app.py", "content": "x = 1"}') === "src/app.py");
ok("file_delete names the file", api.path("file_delete", '{"path": "tmp/x.txt"}') === "tmp/x.txt");
ok("file_delete says deleted", api.op("file_delete") === "deleted");
ok("file_apply_patch says patched", api.op("file_apply_patch") === "patched");
ok("file_write says wrote", api.op("file_write") === "wrote");

process.exit(fail ? 1 : 0);
